python/ray/data/dataset.py

```diff
@@ -4563,7 +4568,37 @@ def stats(self) -> str:
     def _get_stats_summary(self) -> DatasetStatsSummary:
         return self._plan.stats_summary()
 
-    @ConsumptionAPI(pattern="Time complexity:")
+    @ConsumptionAPI(pattern="Examples:")
+    @DeveloperAPI
+    def iter_internal_ref_bundles(self) -> Iterator[RefBundle]:
+        """Get an iterator over ``RefBundles``
+        belonging to this Dataset. Calling this function doesn't keep
+        the data materialized in-memory.
+
+        Examples:
+            >>> import ray
+            >>> ds = ray.data.range(1)
+            >>> for ref_bundle in ds.iter_internal_ref_bundles():
+            ...     for block_ref, block_md in ref_bundle.blocks:
+            ...         block = ray.get(block_ref)
+
+        Returns:
+            An iterator over this Dataset's ``RefBundles``.
+        """
+
+        def _build_ref_bundle(
+            blocks: Tuple[ObjectRef[Block], BlockMetadata],
+        ) -> RefBundle:
+            # Set `owns_blocks=True` so we can destroy the blocks eagerly
+            # after getting count from metadata.
```

Comment:
* Let's also update `_plan.execute_to_iterator` to return RefBundles? 
* Eager free won't work here, unless we explicitly call `destroy_if_owned` after count. I think it's fine to not eager free, just a note about the comment.

Reply:
> Let's also update _plan.execute_to_iterator to return RefBundles?

i am thinking i can do that in a future PR, while i also replace `get_internal_block_refs` usages with the new `iter_internal_ref_bundles` method. what do you think?

> Eager free won't work here, unless we explicitly call destroy_if_owned after count. I think it's fine to not eager free, just a note about the comment.

ah thanks, i had misunderstood how that worked. i will remove the comment but keep `owns_blocks=True` usage

Reply:
> Let's also update _plan.execute_to_iterator to return RefBundles?

@scottjlee @raulchen I'll do this. It was planned clean up item from the LazyBlockList removal